src/lib/sw_engine/tvgSwRaster.cpp

```diff
@@ -142,141 +142,104 @@ static bool _rasterTranslucentRle(SwSurface* surface, SwRleData* rle, uint32_t c
-            if (dst && img && img[index]) {
-                auto alpha = ALPHA_MULTIPLY(span->coverage, opacity);
-                auto src = ALPHA_BLEND(img[index], alpha);
-                auto invAlpha = 255 - surface->comp.alpha(src);
-                *dst = src + ALPHA_BLEND(*dst, invAlpha);
-            }
+            auto alpha = ALPHA_MULTIPLY(span->coverage, opacity);
+            auto src = ALPHA_BLEND(img[rY * w + rX], alpha);     //TODO: need to use image's stride
+            *dst = src + ALPHA_BLEND(*dst, 255 - surface->comp.alpha(src));
+            ++dst;
         }
         ++span;
     }
     return true;
 }
 
 
-static bool _rasterImageRle(SwSurface* surface, SwRleData* rle, uint32_t *img, const Matrix* invTransform, uint32_t w, uint32_t h)
+static bool _rasterImageRle(SwSurface* surface, SwRleData* rle, uint32_t *img, uint32_t w, uint32_t h, const Matrix* invTransform)
 {
-    if (!rle) return false;
-
     auto span = rle->spans;
 
     for (uint32_t i = 0; i < rle->size; ++i) {
+        auto ey1 = span->y * invTransform->e12 + invTransform->e13;
+        auto ey2 = span->y * invTransform->e22 + invTransform->e23;
+        auto dst = &surface->buffer[span->y * surface->stride + span->x];
         for (uint32_t x = 0; x < span->len; ++x) {
-            auto rX = static_cast<uint32_t>(roundf((span->x + x) * invTransform->e11 + span->y * invTransform->e12 + invTransform->e13));
-            auto rY = static_cast<uint32_t>(roundf((span->x + x) * invTransform->e21 + span->y * invTransform->e22 + invTransform->e23));
-
+            auto rX = static_cast<uint32_t>(roundf((span->x + x) * invTransform->e11 + ey1));
+            auto rY = static_cast<uint32_t>(roundf((span->x + x) * invTransform->e21 + ey2));
             if (rX < 0 || rX >= w || rY < 0 || rY >= h) continue;
-
-            auto dst = &surface->buffer[span->y * surface->stride + span->x + x];
-            auto index = rY * w + rX;        //TODO: need to use image's stride
-            if (dst && img && img[index]) {
-                auto src = ALPHA_BLEND(img[index], span->coverage);
-                auto invAlpha = 255 - surface->comp.alpha(src);
-                *dst = src + ALPHA_BLEND(*dst, invAlpha);
-            }
+            auto src = ALPHA_BLEND(img[rY * w + rX], span->coverage);    //TODO: need to use image's stride
+            *dst = src + ALPHA_BLEND(*dst, 255 - surface->comp.alpha(src));
+            ++dst;
         }
         ++span;
     }
     return true;
 }
 
 
-static bool _rasterTranslucentImage(SwSurface* surface, uint32_t *img, uint32_t opacity,const SwBBox& region, const Matrix* invTransform, uint32_t w, uint32_t h)
+static bool _rasterTranslucentImage(SwSurface* surface, uint32_t *img, uint32_t w, uint32_t h, uint32_t opacity, const SwBBox& region, const Matrix* invTransform)
 {
     for (auto y = region.min.y; y < region.max.y; y++) {
+        auto dst = &surface->buffer[y * surface->stride + region.min.x];
```

Comment:
'+x' is missing from the original source.
Could you check this?

Reply:
missed this case:
if (rX < 0 || rX >= w || rY < 0 || rY >= h) continue;